src/BinWrapper.jl

```diff
@@ -0,0 +1,161 @@
+    # create pipes/buffers for in/out/err of a command
+    cmdin  = IOBuffer(input)
+    cmdout = IOBuffer()
+    cmderr = IOBuffer()
+
+    if VERSION < v"1.1"
+      cmdin  = Pipe()
+      cmdout = Pipe()
+      cmderr = Pipe()
+    end
+
+    # run a process
+    process = run(pipeline(ignorestatus(cmd), stdin=cmdin, stdout=cmdout, stderr=cmderr), wait=!(VERSION < v"1.1"))
+
+    if VERSION < v"1.1"
+      # close unnecessary communication tunnels of pipes initialized by pipeline()
+      close(cmdout.in)
+      close(cmderr.in)
+
+      # pass the input to the process and close the related pipe
+      # note: this could be also `write(process, input)`, but then `cmdin` would not be used which is not cool
+      write(cmdin, input)
+      close(cmdin)
+
+      # asynchronously read the outputs of the command
+      out = @async String(read(cmdout))
+      err = @async String(read(cmderr))
+
+      # return all the needed data
+      return (
+        out     = fetch(out),
+        err     = fetch(err),
+        ret     = process.exitcode
+      )
+    end # if VERSION < v"1.1"
+
+    # read the outputs of the command
+    out = String(take!(cmdout))
+    err = String(take!(cmderr))
+
+    # return all the needed data
+    return (
+      out     = out,
+      err     = err,
+      ret     = process.exitcode
+    )
+  end
+end
+
+# This function wraps the binaries that come with the original kbmag library.
+# Here, all the parameters are passed as Strings
+function kbmag_bin_wrapper(program::String; params = (), input::String = "", force_dir::String = ".")
+  # remember the current working directory
+  call_dir = pwd()
+
+  # change the directory if necessary
+  # WARNING: when calling a kbmag binary we need to make sure that
+  # all the input files are in the current working directory
+  if force_dir != "."
+    cd(force_dir)
+  end
+
+  result = low_level_exec(`$kbmag_bin_dir$program $(params)`; input = input)
+
+  # change the current working directory back to the previous one
+  if force_dir != "."
+    cd(call_dir)
+  end
+
+  return result # (out, err, ret)
+end
+
+# This function wraps a kbprog binary
+#     groupname::String   --- input file name
+#     dir::String         --- working directory
+function kbprog_call(groupname::String;
+                     dir::String = kbmag_data_dir)
+
+  res = kbmag_bin_wrapper("kbprog"; params = (groupname, ), force_dir = dir)
+
+  # assure that the call was successful
+  res.ret == 0 || error(res.err)
+
+  # return the whole result
+  return res
+end
+
+# This function wraps a kbprog binary
+#     groupname::String   --- input files basename
+#     words               --- a tuple of input strings
+#     dir::String         --- working directory
+function wordreduce_call(groupname::String,
+                         words;
+                         dir = kbmag_data_dir)
+
+  # check whether input files exist and if not, call kbprog
+  if !isfile(joinpath(dir, groupname * ".kbprog")) ||
```

Comment:
previously it worked because here we have alternative of negations ;-) which is objectively quite hard to reason about.
maybe:
```julia
files = [....]
if !all(isfile.(files))
    ...
```
?? (I wish there were `exists` function:)

Reply:
Thanks for this advice! By the way, `exists` would be hard to define because it is hard to decide what exactly _to exist_ means (in the generic filesystem). Does a pipe exist? Does a symbolic link to a non-existing element exist (as a link) or not (as the linked file)?

Reply:
I meant `exists` as a tongue in cheek as this would read very mathmatically:
`!all(exists.(...)) && ... `  = if not all (files) exist ... ;-)